### PR TITLE
[StructuralMechanicsApplication] Adding CurveFittingHardening to finite strain solution

### PR DESCRIPTION
**Pull request by loumalouomega**

### CODE REVIEW

File: applications/StructuralMechanicsApplication/tests/cl_test/SmallDeformationPlasticity/small_deformation_plasticity_test_parameters.json [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+                        "file_label"          : "step",
+                        "output_control_type" : "step",
+                        "output_frequency"    : 1,
+                        "body_output"         : true,
+                        "node_output"         : false,
+                        "skin_output"         : false,
+                        "plane_output"        : [],
+            "nodal_results"       : ["DISPLACEMENT"],
+            "gauss_point_results" : ["GREEN_LAGRANGE_STRAIN_TENSOR","CAUCHY_STRESS_TENSOR","UNIAXIAL_STRESS","PLASTIC_DISSIPATION"]
+                    },
```
Comment (frastellini) [inline applications/StructuralMechanicsApplication/tests/cl_test/SmallDeformationPlasticity/small_deformation_plasticity_test_parameters.json:115]: I would like to include here "EQUIVALENT_PLASTIC_STRAIN"

File: applications/StructuralMechanicsApplication/custom_constitutive/constitutive_laws_integrators/generic_finite_strain_constitutive_law_integrator_plasticity.h [hunk 2]
```diff
@@ -114,7 +117,8 @@ class GenericFiniteStrainConstitutiveLawIntegratorPlasticity
         LinearSoftening = 0,
         ExponentialSoftening = 1,
         InitialHardeningExponentialSoftening = 2,
-        PerfectPlasticity = 3
+        PerfectPlasticity = 3,
+        CurveFittingHardening = 4
     };
```
Comment (frastellini) [inline applications/StructuralMechanicsApplication/custom_constitutive/constitutive_laws_integrators/generic_finite_strain_constitutive_law_integrator_plasticity.h:122]: Good!

File: applications/StructuralMechanicsApplication/custom_constitutive/constitutive_laws_integrators/generic_finite_strain_constitutive_law_integrator_plasticity.h [hunk 3]
```diff
@@ -244,9 +260,9 @@ class GenericFiniteStrainConstitutiveLawIntegratorPlasticity
             previous_threshold_indicator = threshold_indicator;
 
             // Calculate plastic parameters
-            threshold_indicator = CalculatePlasticParameters(rPredictiveStressVector, rUniaxialStress, rThreshold, rPlasticDenominator, rYieldSurfaceDerivative, rPlasticPotentialDerivative, rPlasticDissipation, delta_plastic_strain, rValues);
+            threshold_indicator = CalculatePlasticParameters(rPredictiveStressVector, rUniaxialStress, rThreshold, rPlasticDenominator, rYieldSurfaceDerivative, rPlasticPotentialDerivative, rPlasticDissipation, delta_plastic_strain, plastic_strain, rValues);
 
-            if (std::abs(threshold_indicator - previous_threshold_indicator) <= std::abs(1.0e-4 * rThreshold)) { // Has converged
+            if (std::abs((threshold_indicator - previous_threshold_indicator)/rThreshold) < ConvergenceTolerance) { // Has converged
                 break;
```
Comment (frastellini) [inline applications/StructuralMechanicsApplication/custom_constitutive/constitutive_laws_integrators/generic_finite_strain_constitutive_law_integrator_plasticity.h:266]: good!

File: applications/StructuralMechanicsApplication/tests/cl_test/SmallDeformationPlasticity/small_deformation_plasticity_test_parameters.json (hunk 1 above)
Comment (frastellini) [reply applications/StructuralMechanicsApplication/tests/cl_test/SmallDeformationPlasticity/small_deformation_plasticity_test_parameters.json:115]: We should register EQUIVALENT_PLASTIC_STRAIN for post-processing

Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/tests/cl_test/SmallDeformationPlasticity/small_deformation_plasticity_test_parameters.json:115]: I can add it

**Comment by loumalouomega** [comment]
@frastellini added

**Review by AlejandroCornejo — APPROVED** [review]

**Review by frastellini — APPROVED** [review]
Perfect. Thank you.

**Pull request merged**